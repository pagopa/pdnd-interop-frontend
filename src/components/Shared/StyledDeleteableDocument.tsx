import React, { useRef, useState } from 'react'
import { Typography } from '@mui/material'
import { Box } from '@mui/system'
import { Delete as DeleteIcon, ModeEdit as ModeEditIcon } from '@mui/icons-material'
import { EServiceDocumentRead } from '../../../types'
import { useFeedback } from '../../hooks/useFeedback'
import { StyledButton } from './StyledButton'
import { StyledTooltip } from './StyledTooltip'
import { forceReflow } from '../../lib/wait-utils'

type StyledDeleteableDocumentComponentProps = {
  eserviceId: string
  descriptorId: string
  readable: EServiceDocumentRead
  deleteDocument: () => Promise<void>
}

export function StyledDeleteableDocument({
  eserviceId,
  descriptorId,
  readable,
  deleteDocument,
}: StyledDeleteableDocumentComponentProps) {
  const { runAction } = useFeedback()
  const contentEditableRef = useRef<HTMLDivElement>(null)
  const [canEdit, setCanEdit] = useState(false)

  // Cross browser place caret at the end of a contenteditable
  // See: https://stackoverflow.com/a/4238971
  const placeCaretAtEnd = (el: HTMLDivElement) => {
    el.focus()
    if (typeof window.getSelection != 'undefined' && typeof document.createRange != 'undefined') {
      const range = document.createRange()
      range.selectNodeContents(el)
      range.collapse(false)
      const sel = window.getSelection()
      if (!sel) return
      sel.removeAllRanges()
      sel.addRange(range)
    }
  }

  const updateCanEdit = async (e: React.SyntheticEvent) => {
    e.preventDefault()
    const newState = !canEdit
    setCanEdit(newState)

    if (newState) {
      await forceReflow()
      const ref = contentEditableRef.current as HTMLDivElement
      ref.focus()
      placeCaretAtEnd(ref)
    }
  }

  const postDescription = async () => {
    const ref = contentEditableRef.current as HTMLDivElement
    await runAction(
      {
        path: {
          endpoint: 'ESERVICE_VERSION_DRAFT_UPDATE_DOCUMENT_DESCRIPTION',
          endpointParams: { eserviceId, descriptorId, documentId: readable.id },
        },
        config: { data: { description: ref.textContent } },
      },
      { suppressToast: true }
    )
  }

  const onBlur = () => {
    setCanEdit(false)
    postDescription()
  }

  return (
    <Box
      sx={{
        display: 'flex',
        alignItems: 'flex-end',
        justifyContent: 'space-between',
        mb: 2,
        pb: 2,
        borderBottom: 1,
        borderColor: 'divider',
      }}
    >
      <Box sx={{ mr: 4, flexShrink: 1 }}>
        <Typography component="span" fontWeight={700}>
          {readable.name}
        </Typography>
        <br />

        <Box
          ref={contentEditableRef}
          contentEditable={canEdit}
          sx={{ mt: 1, py: 2, px: 1, mx: -1 }}
          onBlur={onBlur}
          dangerouslySetInnerHTML={{ __html: decodeURIComponent(readable.description) }}
        />
      </Box>
      <Box sx={{ ml: 4, flexShrink: 0 }}>
        <StyledTooltip title="Modifica descrizione">
          <StyledButton
            sx={{
              px: 1,
              py: 1,
              bgcolor: canEdit ? 'primary.main' : 'transparent',
              color: canEdit ? 'common.white' : 'primary.main',
            }}
            onClick={updateCanEdit}
          >
            <ModeEditIcon fontSize="small" />
          </StyledButton>
        </StyledTooltip>
        <StyledTooltip title="Cancella documento">
          <StyledButton sx={{ px: 1, py: 1 }} onClick={deleteDocument}>
            <DeleteIcon fontSize="small" />
          </StyledButton>
        </StyledTooltip>
      </Box>
    </Box>
  )
}
