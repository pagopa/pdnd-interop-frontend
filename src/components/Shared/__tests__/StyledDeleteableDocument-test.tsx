import React from 'react'
import renderer from 'react-test-renderer'
import { StyledDeleteableDocument } from '../StyledDeleteableDocument'

describe('Snapshot', () => {
  it('matches', () => {
    const eserviceId = 'dsjfisdjfew3ef3e-je9fje3fe3-ejf39fjfej'
    const descriptorId = 'ijoijeoife2ioejk-j9ejfpe29j-je90jfe209'
    const readable = {
      contentType: '',
      description: 'Lorem ipsum dolor sit amet...',
      id: 'disfojasdofjido-sdjifodjofj-djfisdojf',
      name: 'Il mio documento',
    }
    const deleteDocument = () => Promise.resolve()
    const props = { eserviceId, descriptorId, readable, deleteDocument }

    const component = renderer.create(<StyledDeleteableDocument {...props} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
