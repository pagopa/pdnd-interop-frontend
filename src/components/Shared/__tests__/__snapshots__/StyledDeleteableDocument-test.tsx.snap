// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches 1`] = `
<div
  className="MuiBox-root css-1cwi2w5"
>
  <div
    className="MuiBox-root css-1h2t35s"
  >
    <span
      className="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
    >
      Il mio documento
    </span>
    <br />
    <div
      className="MuiBox-root css-oehp33"
      contentEditable={false}
      dangerouslySetInnerHTML={
        Object {
          "__html": "Lorem ipsum dolor sit amet...",
        }
      }
      onBlur={[Function]}
    />
  </div>
  <div
    className="MuiBox-root css-1qu43cj"
  >
    <button
      aria-label="Modifica descrizione"
      aria-labelledby={null}
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-bsol4t-MuiButtonBase-root-MuiButton-root"
      data-mui-internal-clone-element={true}
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="ModeEditIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
        />
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      aria-label="Cancella documento"
      aria-labelledby={null}
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1q3zsck-MuiButtonBase-root-MuiButton-root"
      data-mui-internal-clone-element={true}
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="DeleteIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
        />
      </svg>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
